### ISSUE TITLE
maya plugin: support export and import of instances

### ISSUE THREAD

**Issue by chadrik** [PixarAnimationStudios/USD#142]
@spiffmon just called to my attention that the maya plugin does not currently support export/import of instances (in issue #141).  I'm making this ticket to track progress.  

do you know of any incompatibilities between Maya's and USD's implementations of instancing which would make this task more difficult?

**Comment by spiffmon** [comment]
Hi @chadrik,
   The only Maya instancing feature that neither USD scene-level instancing, nor the PointInstancer, support, is the ability to have per-instance colorsets or uvs (i.e. non-constant primvars) - we allow only constant-interpolation primvars on instances.  I've never really understood what Maya is trying to do there, since as soon as you allow interpolatable primvars on instances, you need to start doing per-instance dicing in your renderer.  Or maybe I just have too subdivvy a mindset :-)

   Otherwise, USD instancing is a superset of Maya's, allowing instancing of entire hierarchies (which is where you get a much, much bigger win in USD itself), as well as nested instancing.

   There are some choices to make in a Maya I/O implementation, however:

1. What to do on import when you encounter a USD instance whose master is not an xform with a single gprim child 

2. How to encode "the master" in USD on export.  Either you love or hate the fact that in Maya the master is still a live object whose edits propagate to all instances.  In USD you can have a choice.  What we've discussed doing is that for each master, we would make an undefined, root level prim that gets all the attributes defined on the Maya master's xform, and put the shape below it.  Then, in the primary scenegraph tree, everywhere there was either an instance or the master itself, we simple reference the undefined prim, and make each referencing prim instanceable.  This encoding is very efficient, and breaks the link between the (viewport-selectable) obejcts that actually live in the primary scenegraph.  You can still make edits to the undefined prim and its child that will propagate to all instances, of course.

The other way you could go is to follow Maya in making the "master" in the scenegraph be an ordinary prim (although we must export it as an Xform with a shape/gprim child, i.e. refrain from doing the normal xform/shape collapsing we do on export), and then make all the other instances inherit or specialize the "master", depending on how you want overrides on the "master" to propagate to the instances.  I use quotes on "master" here because that prim is not a master in the usd-instancing sense - USD will still compose a master for the instances separately from the prim in the scenegraph, but all the opinions in/on the scenegraph prim will flow into the master, and from there to the instances.

**Comment by chadrik** [comment]
> USD instancing is a superset of Maya's, allowing instancing of entire hierarchies (which is where you get a much, much bigger win in USD itself), as well as nested instancing.

You can instance hierarchies in Maya as well:

```python
import pymel.core as pm
pm.newFile(f=1)
c1 = pm.polyCube()[0]
c2 = pm.polyCube()[0]
g1 = pm.group(c1, c2)
g2 = pm.instance(g1)[0]
```

> There are some choices to make in a Maya I/O implementation, however

I need to take some time to fully understand instancing in USD (been trying to grok the thorough but dense (glossary entry)[http://graphics.pixar.com/usd/docs/USD-Glossary.html#USDGlossary-Instancing]) before I can take this conversation much further, but my initial reaction is that the details of Maya's instancing should not affect how we structure an exported USD file, other than the addition of maya-specific metadata to allow proper round-tripping (i.e. the instance index. index 0 is always the master).  Ultimately, it seems that there should be a global set of best practices that extend to any host application (perhaps with strategic variations depending on context).

Perhaps your response is influenced by the work done on `pxrUsdReferenceAssembly`, whose design requires keeping a USD stage *in sync* with the Maya scene.  In this case using USD to model Maya's behavior could be an advantage, but this is internal representation is not necessarily what you want export.  In a referencing-based system (or pure imports without assemblies), there is no internal USD representation, so modeling Maya's behavior is not necessary.  

Does this make sense or am I missing the point?

A few more questions:

- the only instancing examples in the glossary use `inherits` rather than `reference`.  I believe that we want the latter in our simple export/import case, correct?  
- we are not going to have the time to implement instancing for Maya Assemblies, it's not a priority for us and likely a bit more complicated.   This made me wonder if Pixar would consider switching to "old school" referencing once issue #134 is complete.  It would make maintaining the code much simpler, and I've seen little movement on the part of Autodesk to resolve any of the crippling deficiencies in Assemblies.  I've added a section on the pros/cons of references vs assemblies to #134.

**Comment by jtran56** [comment]
Filed as internal issue #141639.

**Comment by chadrik** [comment]
Just a heads up: @palm is going to be taking care of the import side of this soon. Export was completed awhile back.

**Comment by sunyab** [comment]
Closing this out now that the Maya plugin is now being developed in Autodesk's maya-usd repository. Here's a link to an issue tracking this issue in that repo: https://github.com/Autodesk/maya-usd/issues/310

Thanks!
